File: java/RentalManagerReal.java
package videoprokat;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * Реализация класса RentalManager.
 * Содержит конкретную реализацию всех методов абстрактного класса RentalManager.
 */
public class RentalManagerReal extends RentalManager {

    private static int nextId = 1;

    /**
     * Конструктор по умолчанию.
     * Инициализирует список активных аренд и калькулятор.
     */
    public RentalManagerReal() {
        this.activeRentals = new ArrayList<>();
        this.calculator = new FinancialCalculatorReal();
    }

    @Override
    public Rental createRental(Client client, List<VideoCarrier> items, int days,
                               String rentalDate, String plannedReturnDate) {
        if (client == null || items == null || items.isEmpty() || days <= 0) {
            return null;
        }
        List<VideoCarrier> safeItems = new ArrayList<>(items);
        Rental rental = new RentalReal(nextId++, client, safeItems, rentalDate, plannedReturnDate);
        rental.setCalculatedAmounts(calculator, days);
        for (VideoCarrier item : safeItems) {
            if (item != null) {
                item.markAsRented();
            }
        }
        activeRentals.add(rental);
        return rental;
    }

    public Rental createRentalWithException(Client client, List<VideoCarrier> items, int days,
                                            String rentalDate, String plannedReturnDate) throws RentalException {
        if (client == null) {
            throw new RentalException("Клиент не может быть null");
        }
        if (items == null || items.isEmpty()) {
            throw new RentalException("Список носителей не может быть пустым");
        }
        if (days <= 0) {
            throw new RentalException("Количество дней должно быть положительным: " + days);
        }
        
        for (VideoCarrier item : items) {
            if (item == null) {
                throw new RentalException("Элемент в списке не может быть null");
            }
            if (!item.isAvailable()) {
                throw new RentalException("Носитель " + item.getTitle() + " недоступен для аренды");
            }
        }
        
        try {
            List<VideoCarrier> safeItems = new ArrayList<>(items);
            Rental rental = new RentalReal(nextId++, client, safeItems, rentalDate, plannedReturnDate);
            rental.setCalculatedAmounts(calculator, days);
            for (VideoCarrier item : safeItems) {
                item.markAsRented();
            }
            activeRentals.add(rental);
            return rental;
        } catch (Exception e) {
            throw new RentalException("Ошибка при создании аренды: " + e.getMessage(), e);
        }
    }

    @Override
    public double processReturn(Rental rental, int overdueDays) {
        if (rental == null) {
            return 0.0;
        }
        double fine = calculator.calculateOverdueFine(overdueDays);
        for (VideoCarrier item : rental.getItems()) {
            if (item != null) {
                item.markAsAvailable();
            }
        }
        double total = rental.closeRental(fine);
        Iterator<Rental> iterator = activeRentals.iterator();
        while (iterator.hasNext()) {
            if (iterator.next() == rental) {
                iterator.remove();
                break;
            }
        }
        return total;
    }

    @Override
    public List<Rental> getOverdueRentals(String currentDate) {
        List<Rental> result = new ArrayList<>();
        for (Rental rental : activeRentals) {
            if (rental != null && "active".equals(rental.getStatus())) {
                result.add(rental);
            }
        }
        return result;
    }

    @Override
    public void close() {
        if (activeRentals != null) {
            activeRentals.clear();
        }
        calculator = null;
    }

    public static int getNextRentalId() {
        return nextId;
    }

    public static void resetRentalIdCounter() {
        nextId = 1;
    }

    /**
     * Деконструктор (финализатор).
     * Вызывается перед удалением объекта сборщиком мусора.
     */
    @Override
    protected void finalize() throws Throwable {
        try {
            close();
        } finally {
            super.finalize();
        }
    }
}

File: cpp/RentalManagerReal.cpp
#include "RentalManagerReal.hpp"
#include "RentalReal.hpp"
#include "Client.hpp"
#include "VideoCarrier.hpp"
#include "FinancialCalculatorReal.hpp"
#include <algorithm>

int RentalManagerReal::nextId = 1;

RentalManagerReal::RentalManagerReal()
    : RentalManager()
{
    this->activeRentals.clear();
    this->calculator = new FinancialCalculatorReal();
}

RentalManagerReal::~RentalManagerReal()
{
    close();
}

Rental* RentalManagerReal::createRental(Client* client, const std::vector<VideoCarrier*>& items, int days,
                                         const std::string& rentalDate, const std::string& plannedReturnDate)
{
    if (client == nullptr || items.empty() || days <= 0) {
        return nullptr;
    }

    std::vector<VideoCarrier*> safeItems = items;
    Rental* rental = new RentalReal(nextId++, client, safeItems, rentalDate, plannedReturnDate);
    rental->setCalculatedAmounts(calculator, days);

    for (VideoCarrier* item : safeItems) {
        if (item != nullptr) {
            item->markAsRented();
        }
    }

    activeRentals.push_back(rental);
    return rental;
}

double RentalManagerReal::processReturn(Rental* rental, int overdueDays)
{
    if (rental == nullptr) {
        return 0.0;
    }

    double fine = calculator->calculateOverdueFine(overdueDays);

    for (VideoCarrier* item : rental->getItems()) {
        if (item != nullptr) {
            item->markAsAvailable();
            item->incrementRentals();
        }
    }

    double total = rental->closeRental(fine);

    // Удаляем из активных аренд
    auto it = std::find(activeRentals.begin(), activeRentals.end(), rental);
    if (it != activeRentals.end()) {
        activeRentals.erase(it);
    }

    return total;
}

double RentalManagerReal::processReturnWithDamage(Rental* rental, int overdueDays, 
                                                   VideoCarrier* damagedItem, 
                                                   const std::string& damageType, 
                                                   double damageCompensation)
{
    if (rental == nullptr) {
        return 0.0;
    }

    double fine = calculator->calculateOverdueFine(overdueDays);

    for (VideoCarrier* item : rental->getItems()) {
        if (item != nullptr) {
            if (item == damagedItem) {
                if (damageType == "critical") {
                    item->setStatus("written_off");
                } else if (damageType == "minor") {
                    item->setStatus("maintenance");
                } else {
                    item->markAsAvailable();
                }
            } else {
                item->markAsAvailable();
            }
            item->incrementRentals();
        }
    }

    double total = rental->closeRental(fine + damageCompensation);

    // Удаляем из активных аренд
    auto it = std::find(activeRentals.begin(), activeRentals.end(), rental);
    if (it != activeRentals.end()) {
        activeRentals.erase(it);
    }

    return total;
}

std::vector<Rental*> RentalManagerReal::getOverdueRentals(const std::string& currentDate)
{
    std::vector<Rental*> result;
    for (Rental* rental : activeRentals) {
        if (rental != nullptr && rental->getStatus() == "active") {
            result.push_back(rental);
        }
    }
    return result;
}

std::vector<Rental*> RentalManagerReal::getActiveRentals()
{
    std::vector<Rental*> result;
    result.reserve(activeRentals.size());
    for (Rental* rental : activeRentals) {
        result.push_back(rental);
    }
    return result;
}

void RentalManagerReal::close()
{
    if (activeRentals.size() > 0) {
        for (Rental* rental : activeRentals) {
            delete rental;
        }
        activeRentals.clear();
    }
    if (calculator != nullptr) {
        delete calculator;
        calculator = nullptr;
    }
}



